include/alp/algorithms/backsubstitution.hpp

```diff
@@ -0,0 +1,148 @@
+		 */
+		template<
+			typename D = double,
+			typename View,
+			typename ImfR,
+			typename ImfC,
+			typename Vecx,
+			typename Vecb,
+			typename Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			typename Minus = operators::subtract< D >,
+			typename Divide = operators::divide< D >,
+			std::enable_if_t<
+				is_vector< Vecx >::value &&
+				is_vector< Vecb >::value
+			> * = nullptr
+		>
+		RC backsubstitution(
+			Matrix< D, structures::UpperTriangular, Dense, View, ImfR, ImfC > &A,
+			Vecx &x,
+			Vecb &b,
+			const Ring &ring = Ring(),
+			const Minus &minus = Minus(),
+			const Divide &divide = Divide()
+		) {
+
+			RC rc = SUCCESS;
+
+			if( ( nrows( A ) != size( x ) ) || ( size( b ) != size( x ) ) ) {
+				std::cerr << "Incompatible sizes in trsv.\n";
+				return FAILED;
+			}
+
+			const size_t n = nrows( A );
+
+			for( size_t k = 0; k < n ; ++k ) {
+				Scalar< D > alpha( ring.template getZero< D >() );
+				const size_t i = n - k - 1;
+				//x[i]=(b[i]-A[i,i:].dot(x[i:]))/A[i,i]
+ 				auto A_i = get_view( A, i, utils::range( i, n ) );
+				auto A_ii = get_view( A, i, utils::range( i, i + 1 ) );
+				auto x_i = get_view( x, utils::range( i, i + 1 ) );
+				auto b_i = get_view( b, utils::range( i, i + 1 ) );
+				auto x_i_n = get_view( x, utils::range( i, n ) );
+				rc = rc ? rc : alp::dot( alpha, A_i, alp::conjugate( x_i_n ), ring );
+				rc = rc ? rc : alp::set( x_i, b_i );
+				rc = rc ? rc : alp::foldl( x_i, alpha, minus );
+ 				rc = rc ? rc : alp::set( alpha, Scalar< D >( ring.template getZero< D >() ) );
+ 				rc = rc ? rc : alp::foldl( alpha, A_ii, ring.getAdditiveMonoid() );
+ 				rc = rc ? rc : alp::foldl( x_i, alpha, divide );
+			}
+
+			return rc;
+		}
+
+		template<
+			typename D = double,
+			typename ViewA,
+			typename ImfRA,
+			typename ImfCA,
+			typename StructX,
+			typename ViewX,
+			typename ImfRX,
+			typename ImfCX,
+			typename StructB,
+			typename ViewB,
+			typename ImfRB,
+			typename ImfCB,
+			typename Ring = Semiring< operators::add< D >, operators::mul< D >, identities::zero, identities::one >,
+			typename Minus = operators::subtract< D >,
+			typename Divide = operators::divide< D >
+		>
+		RC backsubstitution(
+			Matrix< D, structures::UpperTriangular, Dense, ViewA, ImfRA, ImfCA > &A,
+			Matrix< D, StructX, Dense, ViewX, ImfRX, ImfCX > &X,
+			Matrix< D, StructB, Dense, ViewB, ImfRB, ImfCB > &B,
+			const Ring &ring = Ring()
+		) {
+
+			RC rc = SUCCESS;
+
+			if (
+				( nrows( X ) != nrows( B ) ) ||
+				( ncols( X ) != ncols( B ) ) ||
+				( ncols( A ) != nrows( X ) )
+			) {
+				std::cerr << "Incompatible sizes in trsm.\n";
+				return FAILED;
+			}
+
+			const size_t m = nrows( X );
+			const size_t n = ncols( X );
+
+			for( size_t i = 0; i < n ; ++i ) {
+				auto x = get_view( X, utils::range( 0, m ), i );
+				auto b = get_view( B, utils::range( 0, m ), i );
+				rc = rc ? rc : algorithms::backsubstitution( A, x, b, ring );
```

Comment:
I think it would be easy from this version of  the algorithm to define a bocked one instead but we can have this as a next PR.

Reply:
Yes, this was my thinking. 
Also, I'm not sure if in the blocked Cholesky we want blocked back-substitution.  Probably not, and I now write this only for use in Cholesky.